tools/report_generators/executors.py

```diff
@@ -0,0 +1,181 @@
+
+# logging.basicConfig(
+#     format="%(asctime)s %(threadName)s: %(message)s",
+#     level=logging.DEBUG,
+#     datefmt="%H:%M:%S"
+# )
+
+
+class ReportExecutor:
+    def __init__(
+        self,
+        *,
+        Reporter,
+        wwwroot,
+        kernel_name="",
+        expiry_hours=2,
+        attempts=3,
+        workers=0,
+        force=False,
+        verbose=False,
+        disable_pbar=False,
+        debug=False,
+    ) -> None:
+        self.Reporter = Reporter
+        self.kernel_name = kernel_name
+        self.wwwroot = wwwroot
+        self.expiry_hours = expiry_hours
+        self.attempts = attempts
+        self.workers = workers
+        self.force = force
+        self.verbose = verbose
+        self.disable_pbar = disable_pbar
+        self.debug = debug
+
+    @property
+    def regions(self):
+        regions_all = MetadataRegion.get_all_as_dataframe()
+
+        if self.Reporter.category == "all-regions":
+            selected_regions = regions_all
+        else:
+            selected_regions = regions_all[
+                regions_all["category"] == self.Reporter.category
+            ]
+
+        #  TODO : Not correct as regions actually returns the regions stored in
+        #  the metadata, not the regions to be analysed. This should be
+        #  clarified later on.
+        # if self.debug:
+        #     selected_regions = selected_regions[:10]
+
+        return selected_regions
+
+    def _create_html_report_single(self, region) -> None:
+        for attempt in range(self.attempts):
+            if self.__stop__.is_set():
+                raise KeyboardInterrupt
+
+            logging.info(f"Processing {region} attempt {attempt}")
+            try:
+                report = self.Reporter(
+                    region, wwwroot=self.wwwroot, verbose=self.verbose
+                )
+
+                recently_generated = (
+                    report.metadata.last_updated_hours_ago() < self.expiry_hours
+                )
+                if recently_generated and not self.force:
+                    break
+
+                report.generate(kernel_name=self.kernel_name)
+                break  #  Without this break if force is on it will keep attempting
+            except Exception as e:
+                if e == KeyboardInterrupt:
+                    raise e
+
+                if attempt + 1 == self.attempts:
+                    logging.warning(f"Processing {region} error\n{e}")
+                    raise e
+
+                logging.warning(
+                    f"Processing {region} error {type(e)}, retrying {attempt+1}"
+                )
+
+    def _create_html_reports_serial(self, regions):
+        if self.disable_pbar:
+            pbar = range(len(regions))
+        else:
+            pbar = tqdm(range(len(regions)))
+
+        try:
+            for i in pbar:
+                region = regions[i]
+                region_str = region[-1] if type(region) == list else region
```

Comment:
Looking back through this, what's done here isn't explained and is not that obvious.

`region` can be a string, e.g. `France` as passed through by the `CountryReport`, in that case it's pretty simple as your region is... just the region.

But in the case of Germany we have regions and subregions, so then `region` is a list of  `[Bundesland, Kreis]`, e.g. `['Hamburg', 'SK Hamburg']`. The convention for this would be large -> small, so the final element of the list is what I take as the name.

This approach allows for some flexibility in the future as you can have multiple levels of subregions.

In hindsight it might be better to remove the option of specifying them as a string and always have the region as a list even if there's only one entry.